### PR TITLE
feat(dashboards): KPI comparison & format polish (#3207)

### LINKED ISSUE #3207: Polish KPI cards — period-over-period delta + value formats + sparkline
The KPI / scorecard card (#3137) shipped the core scaffold — headline number, optional `comparisonSql` delta chip, value formats. This slice deepens the "scorecard" feel from the [#2267 design pass](https://github.com/AtlasDevHQ/atlas/issues/2267) deferred list ("Comparison deltas on KPIs — the math layer could go deeper"): infer period-over-period comparisons so the agent doesn't have to hand-write a second query for the common "vs last period" case, tighten value formatting, and polish the optional sparkline.

## Key files
- `packages/types/src/dashboard.ts` — `DashboardKpiConfig`, `DashboardKpiValueFormat`
- `packages/schemas/src/dashboard.ts` — Zod mirror
- `packages/web/src/ui/components/dashboards/kpi-card.tsx` — delta chip, `formatKpiValue`, sparkline
- `packages/api/src/api/routes/dashboards.ts` — render-endpoint comparison execution (`renderCardRoute`)
- `packages/api/src/lib/dashboards.ts` — `rowToCard` / KPI config mapping

## Acceptance criteria
- [ ] Period-over-period inference: when a KPI card binds the dashboard date params (`:date_from`/`:date_to`), a config flag (or agent surface) can request an automatic prior-period comparison without a hand-written `comparisonSql` — the prior window is derived by shifting the bound range back one period and run through the SAME SQL guard (parameterized, auto-LIMIT, timeout).
- [ ] `formatKpiValue` edge cases: negative deltas, zero baseline (no divide-by-zero → "—"/"n/a"), compact large numbers (`1.2M`), `percent`/`currency`/`duration` paths.
- [ ] Delta chip shows a direction arrow + up/down color with an `inverse` option (lower-is-better metrics like churn render green on a decrease).
- [ ] Optional sparkline renders cleanly for a multi-row trend; hidden for single-row scorecards.
- [ ] Agent guidance updated so KPI cards lead with a comparison where a date param exists.
- [ ] Unit tests for the formatter + the comparison-window derivation.

## Dependencies
- Builds on #3137 (KPI card). No hard blockers.

### PR DESCRIPTION
**Pull request by msywulak**
Closes #3207. Deepens the KPI / scorecard card (#3137) for v0.0.10 so it feels finished — automatic period-over-period comparison, hardened formatting, lower-is-better delta coloring, polished sparkline.

### Period-over-period inference (no hand-written `comparisonSql`)
- New `DashboardKpiConfig.autoComparison`. When set, `renderCardRoute` re-runs the card's **own** SQL with the bound date window shifted back exactly one period, run through the **same** SQL guard (parameterized bind, auto-LIMIT, statement timeout, RLS, audit) — never string-interpolated.
- The shift is a pure function: `derivePriorPeriodValues(values, {from, to})` in `lib/dashboard-parameters.ts`. A `[from, to)` window of length N days → the immediately-preceding, non-overlapping N-day window (the half-open convention the docs recommend). Returns `null` (→ comparison skipped, headline still renders) for a missing/unparseable bound or an inverted/empty range.
- Defaults to the `:date_from` / `:date_to` pair; overridable via `comparisonDateParams`. Mutually exclusive with `comparisonSql` (Zod-enforced).

### Delta chip `inverse` (lower-is-better)
- `inverse: true` flips the chip **colour** so a *decrease* reads green (churn, latency, error rate, cost). The direction arrow still tracks the real change; only the tone flips. Pure `deltaTone(direction, inverse)`.

### Formatter + sparkline hardening
- `formatKpiValue` edge cases locked in by tests: negative deltas, zero baseline (divide-by-zero → no chip), compact large numbers (`1.2M`/`2.4B`), `percent`/`currency`/`duration`, non-finite → `—`.
- `sparklineGeometry` (pure, tested): a **flat** series now centres vertically instead of pinning to the bottom edge; an end-of-line dot marks the latest point; the line is tinted by delta tone; hidden for single-row scorecards.

### Agent guidance
- `createDashboard` now leads with `autoComparison` where date params exist, documents `inverse`, and **rejects** `autoComparison` on a card whose SQL doesn't filter by the window (otherwise the prior-period query equals the primary — a no-op delta).

### Wiring
- `hasKpiComparison` / `kpiComparisonSignature` widened so autoComparison cards fetch their delta on **first load** (not only on param change). The client-only `inverse` colour is excluded from the fetch signature (toggling it must not refetch).

## Back-compatible
A KPI card with no comparison config renders exactly as today. `rowToCard` carries the new fields through unchanged (the `chart_config` blob is parsed whole — no field-by-field mapping strips them); covered by a round-trip test.

## Versioning
- `@useatlas/types` `0.1.15 → 0.1.16` (additive `DashboardKpiConfig` fields), following the #3137 precedent. `@useatlas/schemas` (source-bundled, `workspace:*`) not bumped. `openapi.json` regenerated for the new fields in the Add/Update-card route bodies.

## Tests / gates
TDD throughout (formatter + window derivation are pure functions; red tests first). New unit tests in: `dashboard-parameters.test.ts` (window derivation, 13 cases), `schemas/dashboard.test.ts` (autoComparison/inverse validation), `kpi-card.test.tsx` (deltaTone, sparklineGeometry, inverse rendering, widened signature), `dashboards.test.ts` (route autoComparison), `create-dashboard.test.ts` (autoComparison validation), `dashboards-row-to-card.test.ts` (round-trip).

Local `/ci` green: lint ✓ · type ✓ · full `bun run test` ✓ · syncpack ✓ · template-drift ✓ · openapi-drift regenerated ✓.

> Shares `types/src/dashboard.ts` + `schemas/src/dashboard.ts` with the #3212 drilldown session — different regions (KPI interfaces vs drilldown config). Rebased on latest `origin/main` (unchanged at branch time).

---
<a href="https://app.blacksmith.sh/AtlasDevHQ/codesmith/atlas/pr/3215"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1783218561&installation_id=135337507&pr_number=3215&repository=AtlasDevHQ%2Fatlas&return_to=https%3A%2F%2Fgithub.com%2FAtlasDevHQ%2Fatlas%2Fpull%2F3215&signature=8a963dd0a1bed466c7c5c084bca62853ae3ca05c295010b37135700f8f9e1bc6"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * KPI cards can auto-calculate period-over-period comparisons by re-running their query against a shifted prior-date window (configurable date-parameter keys).
  * Delta chips gain an `inverse` option to flip color semantics; tone maps separately from arrow direction.
  * Improved sparkline geometry and rendering; comparisons re-run when date parameters change.

* **Bug Fixes**
  * Misconfigured auto-comparison is rejected at card save to prevent broken KPIs.
  * When a prior window can’t be derived, comparison is omitted gracefully (no render failure).

### CODE REVIEW

File: packages/web/src/ui/components/dashboards/kpi-card.tsx [hunk 1]
```diff
@@ -112,17 +129,35 @@ export function formatKpiValue(value: number | null, format?: DashboardKpiValueF
...
+/** The fetch-affecting comparison config of a KPI card — what the `/render`
+ *  endpoint actually runs: the hand-written query, or the auto period-over-period
+ *  window. The client-only `inverse` (colour) is excluded: it changes how the
+ *  delta is painted, not what's fetched, so toggling it must NOT refetch. */
+function comparisonKey(card: Pick<DashboardCard, "chartConfig">): unknown {
+  const kpi = card.chartConfig?.kpi;
+  return {
+    sql: kpi?.comparisonSql ?? "",
+    auto: kpi?.autoComparison === true,
+    params: kpi?.comparisonDateParams ?? null,
```
[the remark this answers is not in the document]
Comment (msywulak) [reply packages/web/src/ui/components/dashboards/kpi-card.tsx:158]: Good catch — fixed in 3e897b52. `kpiComparisonSignature` now folds the card's primary `sql` into the key for auto cards (`autoSql: kpi.autoComparison ? card.sql : ""`), so editing the primary SQL while `autoComparison` stays on moves the signature and re-fires `loadDefaultComparisons()`. `inverse` stays out of the key (display-only). Added tests: *moves when an autoComparison card's primary SQL changes* and *a comparisonSql card's signature ignores its primary SQL*.

File: packages/schemas/src/dashboard.ts [hunk 2]
```diff
@@ -209,28 +209,76 @@ export type DashboardKpiValueFormatWire = z.infer<typeof dashboardKpiValueFormat
...
 export const dashboardKpiConfigSchema = z
   .object({
     valueFormat: dashboardKpiValueFormatSchema.optional(),
     comparisonSql: z.string().min(1).max(DASHBOARD_KPI_COMPARISON_SQL_MAX).optional(),
+    /**
+     * #3207 — request an automatic prior-period comparison instead of a
+     * hand-written `comparisonSql`. The render endpoint re-runs the card's own
+     * SQL with the bound date window shifted back one period.
+     */
+    autoComparison: z.boolean().optional(),
```
[the remark this answers is not in the document]
Comment (msywulak) [reply packages/schemas/src/dashboard.ts:235]: Agreed — the window-reference check is now shared across every persistence path. Extracted `validateAutoComparison(sql, kpi, parameters?)` in `lib/dashboard-parameters.ts` and wired it into the REST add/update-card routes and the bound-editor add/update tools, in addition to createDashboard (3e897b52). The schema itself can't see the card SQL, so per-path validation is the enforcement point; the render path still degrades gracefully (omits the delta + logs) as a backstop. The bound editor runs the reference check (it has the card SQL only); the date-type check runs where the dashboard's parameter defs are loaded (createDashboard + REST routes).

File: packages/api/src/lib/tools/create-dashboard.ts [hunk 3]
```diff
@@ -312,6 +314,26 @@ If any card has invalid SQL or references an undeclared parameter, the whole cal
...
+        // period and re-runs the SAME sql. That only produces a meaningful
+        // delta if the card actually FILTERS by both window params — otherwise
+        // the prior-period query is identical to the primary (a no-op delta).
+        // (The window params being undeclared is already caught above, since a
+        // referenced-but-undeclared param fails the check; this catches the
+        // distinct "autoComparison but the SQL ignores the window" mistake.)
+        const kpi = card.chartConfig.kpi;
+        if (kpi?.autoComparison) {
+          const { from, to } = kpi.comparisonDateParams ?? DEFAULT_COMPARISON_DATE_PARAMS;
+          const missingRef = [from, to].filter((name) => !referenced.has(name));
```
[the remark this answers is not in the document]
Comment (msywulak) [reply packages/api/src/lib/tools/create-dashboard.ts]: Fixed in 3e897b52. `validateAutoComparison` now also checks that both window params are declared `type: "date"` (where the dashboard's parameter defs are available — createDashboard + the REST routes). A `number`/`text` param is rejected with an actionable message instead of silently producing a null prior-period window. Added a createDashboard test (*rejects autoComparison when a window param is not declared as a date*) plus full-branch unit tests for the helper.

**Comment by msywulak** [comment]
Pushed 3e897b52 addressing all three Codex P2 findings (auto-comparison signature now includes the primary SQL; shared `validateAutoComparison` enforced on the REST add/update-card routes + bound editor; date-typed window-param check). Replied inline on each thread.

@codex review
@coderabbitai review
